gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/base.py.j2

```diff
@@ -80,6 +95,21 @@ class {{ service.name }}Transport(abc.ABC):
             host += ':443'
         self._host = host
 
+        # If a custom API endpoint is set, set scopes to ensure the auth
+        # library does not used the self-signed JWT flow for service
+        # accounts
+        if host != self.DEFAULT_HOST and not scopes:
+            scopes = self.AUTH_SCOPES
```

Comment:
If port is provided (say `pubsub.googleapis.com:443` instead of `pubsub.googleapis.com`) the library will think the OAuth2 flow is necessary.

I initially tried splitting on `:` but that broke the showcase tests where default host _is_ localhost.  Is there a better way to parse the provided host reliably?

Reply:
Silly questions: which is the OAuth2 flow, and why do we not want it all the time? Is it the "old style"? What are we trying to parse out? We can always fall back to regexes, something like

host_re = re.compile("(?P<host>[^:]+)(:(?P<port>\d+))?")

Or is that not the part of parsing out the host that's problematic?

Reply:
Ah sorry for being vague.

The OAuth2 flow (currently always used for service accounts) requires a call out to `oauth2.googleapis.com`. With a service account, it's sometimes possible to "self-sign" the token, which avoids a network round trip.

The self-signed JWT option can be used when:

1. A service account is being used
2. A scope has not been set by the user
3. A target_audience has not been set by the user (irrelevant as we don't have an client option for this)
4. An api_endpoint has not been set by the user

Number 4 is what's relevant here. If the user is using the non-default endpoint we want to make sure the auth library uses OAuth2. The suggested way to do this in the [doc](https://docs.google.com/document/d/1SNCVTmW6Rtr__u-_V7nsT9PhSzjj1z0P9fAD3YUgRoc/edit#heading=h.r0s8lehl5drr) is to pass user scopes to the auth library.

I think that regex will work! I'll experiment and push a new commit.